### ISSUE TITLE
SCIM sync not updating `active` attribute on deactivation

### ISSUE THREAD

**Issue by terussell85** [ssoready/ssoready#155]
While testing SCIM, I noticed that deleted/deactivated users did not get their `active` attribute updated, despite having been marked as deleted in SSOReady.

I'm not a SAML expert by any means, but based on what I read, I think that the behavior is slightly wrong.

**Reproduction Steps:**
1. Setup a SAML app and SCIM sync to Entra ID
1. Create a user in Entra ID
1. Manually sync the new user from Entra ID
1. Mark the user's account as "not enabled" in Entra ID (uncheck the Account Enabled property). This is not the same as deletion.
1. Manually sync again from Entra ID
1. Observe the PATCH request in SSOReady
Should have the following body:
```
{
    "Operations": [
        {
            "op": "Replace",
            "path": "active",
            "value": "False"
        }
    ],
    "schemas": [
        "urn:ietf:params:scim:api:messages:2.0:PatchOp"
    ]
}
```

**Actual Results**
Examine the user attributes in SSOReady. 
Notice that the deleted property is correctly set to `true`, but that the `active` attribute remains as `true`.

**Expected Results**
The deleted property would be set to `true` and the `active` attribute would be set to `false`.

**Comment by ucarion** [comment]
Just to be clear, you're not reporting that the SSOReady SCIM User has an incorrect `deleted`, right? Just that the `attributes["active"]` is "stale" in the face of deprovisioning? If so, this is a bit of a known problem, and I agree it's confusing.

If you want immediate advice on what to do here, my recommendation is to ignore `attributes["active"]` on SCIM users. SSOReady special-cases `active`, and the `deleted` property on a SCIM user is much better-behaved than `attributes["active"]`.

The longer version of what's up here:

What's going here has to do with working around Entra's noncompliance with SCIM. The `attributes` on an SSOReady user try to be a "dumb" record of what we know about a SCIM user. This works in the obvious way for everything _except_ `active`, because:

1. Deleting a SCIM user (i.e. by doing a SCIM `DELETE /Users/...`) doesn't technically touch any attribute. That's part of why SSOReady tracks `deleted` as a special property.
2. Microsoft in particular is a poor actor here. At the risk of telling you what you already know, their default behavior (which you can override, but it's pretty esoteric how you do this) for deprovisioning is to set the SCIM `active` attribute to the JSON string `"True"` or `"False"`. That's straight-up against the SCIM spec, which requires it to be a JSON boolean `true` or `false`.

Notice that in your example, the Entra payload is:

```json
        {
            "op": "Replace",
            "path": "active",
            "value": "False"
        }
```

The SCIM spec requires, in contrast to Entra's real-world behavior, that the payload instead be:

```json
        {
            "op": "Replace",
            "path": "active",
            "value": false
        }
```

That is: `false` instead of `"False"`. That's what requires SSOReady to rewrite requests that come from Entra. That rewriting process also drops the `active` property from `PATCH` requests. It's not super clear to me what the desired behavior _should_ be -- I think every option is confusing here.

Very open to thoughts here. I consider the "bad vibes" of weirdness around `properties["active"]` to be a bug.

**Comment by terussell85** [comment]
> Just to be clear, you're not reporting that the SSOReady SCIM User has an incorrect deleted, right? Just that the attributes["active"] is "stale" in the face of deprovisioning? If so, this is a bit of a known problem, and I agree it's confusing.

That's correct. The `deleted` property is accurate.

> If you want immediate advice on what to do here, my recommendation is to ignore attributes["active"] on SCIM users. SSOReady special-cases active, and the deleted property on a SCIM user is much better-behaved than attributes["active"].

👍 That checks out and matches my interpretation based on what I was observing as well. My report was mostly raised under the thought of "Oh, I wonder if they've noticed this" because it _felt_ wrong when I saw it.

The only use case I _might_ see as valid is if I need to show in my app whether a user was temporarily disabled or if it was deleted. Right now both are boiled down to the same `deleted` flag. But that's a stretch. I wouldn't recommended any updates based solely on that.

> It's not super clear to me what the desired behavior should be -- I think every option is confusing here.
I see your point. I didn't catch the string "False" value. Looks like a stereotypical serialization issue on their side.

The part I think I'm missing is why the Entra request rewriter drops the `active` property.  As a user it's hard to separate my consumer mentality from my developer mentality, so I'll just write them separately. 😆 

As a consumer, I would expect active to be set to `false`. Even if Entra is non-conformant, the _intent_ is clear from their request. 

As a developer, I wouldn't want to open the flood gates of custom SCIM logic per-provider and I would want to stick to the SCIM spec as strictly as possible. I think that there is a bigger problem with the code today though. If I'm reading the code correctly, it's a one-off case only for deletion. https://github.com/ssoready/ssoready/blob/c8fc6b74ee80543a3bd45da2f76a8116c6ed20c1/internal/authservice/scim.go#L387 correctly, 
What if that same active gets written back to "True"? I'd expect that to "undelete" the SCIM user.

Sure enough, I ran a quick test and it looks like a manual SCIM provision from Entra ID won't even run on a user that was marked as disabled and then re-enabled. The GET call to `/v1/scim/<directory>/Users/<user>" fails. In fact, oddly enough it seems to fail with a 502 Bad Gateway on your hosted service. I'm guessing a downstream service is responding with 404 and then it's getting modified somewhere up the chain.

**Comment by terussell85** [comment]
After re-reading your first comment a little more... 

is the crux of the problem that you can't tell the difference between "Account Disable" and "Account Delete" actions from Entra? Both are look the same?

So when things are "deleted" (soft deleted) on the SSOReady side, you wouldn't be able to tell if something should be allowed to be undeleted (Account Enable) vs should not be allowed to be undeleted (Account Deleted. Should create new account)?

**Comment by terussell85** [comment]
After re-reading your first comment a little more... I think I'm understanding the predicament a bit better.

Unfortunately, I think there likely may need to be some sort of `enabled` concept versus strictly having a `deleted` concept. If `deleted` is true, `enabled` should always be true, but its possible that `enabled` could be temporarily false while `deleted` is false. For example, accounts that are temporarily disabled for things like personal leave. 

I wouldn't want to complicate things, but I think your options are:
1. More state like `enabled`
2. Allowing the SCIM endpoints to return deleted users, which I'm betting is against the spec.

In those cases, downstream customers will want to be able to enable/disable accounts and have those synced with SCIM. If SSOReady treats enable/disable as delete, then the SCIM endpoints would fail for retrieving existing users, which is likely what I'm seeing above.

That said, I haven't been able to isolate exactly what Entra does in the case of DELETE (versus disable). If it's the same PATCH operation, then I'm not sure what options even exist.

**Comment by ucarion** [comment]
> Sure enough, I ran a quick test and it looks like a manual SCIM provision from Entra ID won't even run on a user that was marked as disabled and then re-enabled. Unfortunately it doesn't even get to the PATCH call because the GET call to /v1/scim/<directory>/Users/<user> fails. In fact, oddly enough it seems to fail with a 502 Bad Gateway on your hosted service. I'm guessing a downstream service is responding with 404 and then it's getting modified somewhere up the chain.

This sounds like a bug. I'm investigating this.

**Comment by ucarion** [comment]
Separately: I would advise against attempting to distinguish a "soft-delete" (i.e. SCIM PATCH/PUT `active` to `false`) from a hard-delete (i.e. SCIM DELETE). Identity providers are not consistent.

**Comment by terussell85** [comment]
> Separately: I would advise against attempting to distinguish a "soft-delete" (i.e. SCIM PATCH/PUT active to false) from a hard-delete (i.e. SCIM DELETE). Identity providers are not consistent.

That's fair. Maybe it's all a null point if the `/v1/scim/<directory>/Users/<user>` is a bug. I'll hold until there is more information. Thanks

**Comment by terussell85** [comment]
I'm closing this issue. With the fixes provided in #157, #158, and #159 the SCIM enable/disable path appears to work fine with Entra now. The `active` flag showing one way or the other is much less important and mostly irrelevant for my needs now that the flow is working.
